diff -r -u insight-6.8.orig/tcl/win/tclWin32Dll.c insight-6.8/tcl/win/tclWin32Dll.c
--- insight-6.8.orig/tcl/win/tclWin32Dll.c	2008-03-04 00:26:42.000000000 +0100
+++ insight-6.8/tcl/win/tclWin32Dll.c	2008-10-14 10:10:53.645535900 +0200
@@ -142,7 +142,7 @@
 };
 
 TclWinProcs *tclWinProcs;
-static Tcl_Encoding tclWinTCharEncoding;
+Tcl_Encoding tclWinTCharEncoding;
 
 /*
  * The following declaration is for the VC++ DLL entry point.
